apt-hook/json-hook.cc

```diff
@@ -0,0 +1,302 @@
+#include <fstream>
+#include <iostream>
+#include <json-c/json.h>
+#include <sstream>
+#include <string>
+#include <vector>
+
+struct jsonrpc_request {
+    bool notification;
+    int64_t id;
+    json_object *root_msg;
+    std::string method;
+    json_object *params;
+};
+
+bool read_jsonrpc_request(std::istream &in, jsonrpc_request &req) {
+    std::string msg_line;
+    std::string empty_line;
+    getline(in, msg_line);
+    getline(in, empty_line);
+    json_object *msg = json_tokener_parse(msg_line.c_str());
+    json_object *tmp;
+    bool has_key = false;
+
+    has_key = json_object_object_get_ex(msg, "jsonrpc", &tmp);
+    if (!has_key) {
+        json_object_put(msg);
+        return false;
+    }
+    std::string msg_jsonrpc_version(json_object_get_string(tmp));
+    if (msg_jsonrpc_version != "2.0") {
+        json_object_put(msg);
+        return false;
+    }
+
+    has_key = json_object_object_get_ex(msg, "method", &tmp);
+    if (!has_key) {
+        json_object_put(msg);
+        return false;
+    }
+    std::string msg_method(json_object_get_string(tmp));
+    req.method = msg_method;
+
+    has_key = json_object_object_get_ex(msg, "params", &tmp);
+    if (!has_key) {
+        json_object_put(msg);
+        return false;
+    }
+    req.params = tmp;
+
+    has_key = json_object_object_get_ex(msg, "id", &tmp);
+    if (has_key) {
+        req.id = json_object_get_int64(tmp);
+        req.notification = false;
+    } else {
+        req.notification = true;
+    }
+
+    req.root_msg = msg;
+
+    return true;
+}
+
+bool string_ends_with(std::string str, std::string ends_with) {
+    if (str.length() >= ends_with.length()) {
+        int result = ends_with.compare(0, ends_with.length(), str, str.length() - ends_with.length(), ends_with.length());
+        return result == 0;
+    }
+    return false;
+}
+
+bool version_from_origin_and_archive_ends_with(json_object *version, std::string from_origin, std::string archive_ends_with) {
+    bool has_key = false;
+    json_object *origins;
+    has_key = json_object_object_get_ex(version, "origins", &origins);
+    if (!has_key) {
+        return false;
+    }
+    int64_t origins_length = json_object_array_length(origins);
+
+    for (int64_t i = 0; i < origins_length; i++) {
+        json_object *origin = json_object_array_get_idx(origins, i);
+
+        json_object *tmp;
+        has_key = json_object_object_get_ex(origin, "origin", &tmp);
+        if (!has_key) {
+            continue;
+        }
+        std::string origin_origin(json_object_get_string(tmp));
+        has_key = json_object_object_get_ex(origin, "archive", &tmp);
+        if (!has_key) {
+            continue;
+        }
+        std::string origin_archive(json_object_get_string(tmp));
+
+        if (origin_origin == from_origin && string_ends_with(origin_archive, archive_ends_with)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+struct security_package_counts {
```

Comment:
We could use a class here and initialize the count variables

Reply:
We could :) do you think it is needed here? I feel that the use-case for this struct is simple enough and we're initializing the variables in the one place it matters so it's okay. But I could be convinced otherwise :)